### ISSUE TITLE
Multiple `Iterator` generic parameters confuse the type inference

### ISSUE THREAD

**Issue by TDplai** [rust-lang/rust-analyzer#15145]
rust-analyzer gets confused about `IntoIterator::Item` types when multiple generic parameters implement `IntoIterator`.

Example:

```rust
pub fn item_type_confusion<'a, I: IntoIterator<Item = &'a mut usize>, I2: IntoIterator>(it: I) {
    for a in it {
        // rust-analyzer reports error: cannot mutate immutable variable `a`
        *a += 1;
    }
}
```

This error is only reported in rust-analyzer. The function compiles without errors. Furthermore, neither `cargo check` nor `cargo clippy` report any errors.

The hover-text on `a` is `a: {unknown}`, which shows that rust-analyzer was unable to infer the type of `a`.

This occurs even when `I` and `I2` have the same `Item` type:

```rust
pub fn analyzer_confusion<
    'a,
    I: IntoIterator<Item = &'a mut usize>,
    I2: IntoIterator<Item = &'a mut usize>,
>(
    i1: I,
) {
    for a in i1 {
        // rust-analyzer reports error: cannot mutate immutable variable `a`
        *a += 1;
    }
}
```

If I coerce  `a` into an `&mut usize`, the error goes away:

```rust
pub fn item_type_confusion<'a, I: IntoIterator<Item = &'a mut usize>, I2: IntoIterator>(it: I) {
    for a in it {
        let _: &mut usize = a;
        // no error reported
        *a += 1;
    }
}
```

rust-analyzer does recognise that `a` is an `&mut usize` when printing a type error, but still claims that `*a` is immutable:

```rust
pub fn item_type_confusion<'a, I: IntoIterator<Item = &'a mut usize>, I2: IntoIterator>(it: I) {
    for a in it {
        // rust-analyzer reports the expected error: expected `u64`, found `&mut usize`
        let _: u64 = a;
        // rust-analyzer reports error: cannot mutate immutable variable `a`
        *a += 1;
    }
}
```

Furthermore, if `I2: IntoIterator` is removed, the error goes away:

```rust
pub fn item_type_confusion<'a, I: IntoIterator<Item = &'a mut usize>, I2>(it: I) {
    for a in it {
        // no error reported
        *a += 1;
    }
}
```

**rust-analyzer version**: rust-analyzer 1.72.0-nightly (8084f39 2023-06-25)

**rustc version**: rustc 1.72.0-nightly (8084f397c 2023-06-25)

**relevant settings**:
`CARGO_HOME` is set to `$HOME/.local/share/cargo`
`RUSTUP_HOME` is set to `$HOME/.local/share/rustup`
`RUSTC_WRAPPER` is set to `sccache`
rust-analyzer settings are:
```lua
["rust-analyzer"] = {
	cargo = {
		features = "all",
	},
	checkOnSave = {
		command = "clippy",
		allFeatures = true,
		extraArgs = { "--", "-W", "clippy::pedantic" },
	},
},
```
(this is copied verbatim from my Neovim configuration)

**Comment by TDplai** [comment]
In further testing, if I coerce `a` into `&mut _`, rust-analyzer is able to infer the `usize` part.

```rust
pub fn analyzer_confusion<'a, I: IntoIterator<Item = &'a mut usize>, I2: IntoIterator>(i1: I) {
    for a in i1 {
        let _: &mut _ = a;
        // no error reported
        *a += 1;
    }
}
```

**Comment by lowr** [comment]
Thanks for the report -- this is a known issue that needs to be solved in an upstream library: rust-lang/chalk#750. Closing as a duplicate of #10653.
